### PR TITLE
calibre: gui to choose default extension

### PR DESCRIPTION
**Pull request by pazos**
Also include sensible defaults (epub as 1st option)

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8146)

### CODE REVIEW

**Comment by pazos** [comment]
I'm not very convinced about this aproach. Probably a more user-friendly way is:

1. get a subtable of `valid_ext` with file formats that calibre can output.
2. show the sortwidget with the subtable
3. send to calibre a table with sorted output + the rest of formats.

That makes the whole thing just a menu entry where you select **one** format as the desired conversion output when sending an unsupported format to the device.

Current implementation allows sorting the entire table, which is mostly useless as we only care about the first format that calibre converts to.

Some context in https://www.mobileread.com/forums/showthread.php?t=341423

**Review by Frenzie — APPROVED** [review]
Sounds a bit overengineered but seems fine in principle. Also it's a bit hard to gauge for me what's required by calibre stuff.

**Comment by pazos** [comment]
> Sounds a bit overengineered

Did you mean the alternative or current implementation?.

File: plugins/calibre.koplugin/extensions.lua [hunk 1]
```diff
@@ -1,37 +1,84 @@
...
+    local SortWidget = require("ui/widget/sortwidget")
+    local _ = require("gettext")
+    local item_table = {}
+    for i = 1, #self.user_preferences do
+        item_table[i] = { text = self.user_preferences[i], label = self.user_preferences[i] }
+    end
+
+    local sort_item
+    sort_item = SortWidget:new{
+        title = _("Sort extensions"),
```
Comment (pazos) [inline plugins/calibre.koplugin/extensions.lua:71]: is name ok?

File: plugins/calibre.koplugin/main.lua [hunk 2]
```diff
@@ -335,6 +337,19 @@ function Calibre:getWirelessMenuTable()
             },
         },
     }
+
+    if not CalibreExtensions:isCustom() then
+        table.insert(t, 2, {
+            text = _("File formats"),
```
Comment (pazos) [inline plugins/calibre.koplugin/main.lua:343]: is name ok?

**Comment by Frenzie** [comment]
@pazos I already approved it, so yes. ;-D

> Did you mean the alternative or current implementation?

I mean being able to sort everything when that apparently doesn't serve a purpose.

**Comment by pazos** [comment]
> I mean being able to sort everything when that apparently doesn't serve a purpose.

Agree. calibre has a list of supported outputs in https://manual.calibre-ebook.com/faq.html

> Output Formats: AZW3, EPUB, DOCX, FB2, HTMLZ, OEB, LIT, LRF, MOBI, PDB, PMLZ, RB, PDF, RTF, SNB, TCR, TXT, TXTZ, ZIP

Discarding formats that we don't support and others that we don't support well the list will become:

> EPUB, DOCX, FB2, MOBI, PDF, TXT

**Comment by mergen3107** [comment]
> the list will become

How about `.fb2.zip` format?

**Comment by pazos** [comment]
> How about .fb2.zip format?

@mergen3107: you tell me :)

AFAIK these are two different extensions: `fb2` and `zip`. Both are declared by us so please try to send a book wirelessly from calibre to your device that's only available as `fb2.zip` on your calibre library and tell me if the book is pushed to your device as `fb2` or as `fb2.zip`. If calibre figures out that none of the supported formats is valid it will convert it to the first format on the list (that's currently `docx`).

**Comment by mergen3107** [comment]
> a book wirelessly from calibre to your device that's only available as `fb2.zip`

Ah, my bad, sorry. Calibre actually does not store `fb2` as `fb2.zip`, only as `fb2`. So my suggestion is invalid.

**Comment by pazos** [comment]
That's better :)

muchbetter

**Review by Frenzie — APPROVED** [review]
lgtm!

**Pull request merged**